group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -434,22 +437,36 @@ public void setSubscriptionMetadata(
     public Map<String, TopicMetadata> computeSubscriptionMetadata(
         ConsumerGroupMember oldMember,
         ConsumerGroupMember newMember,
-        TopicsImage topicsImage
+        TopicsImage topicsImage,
+        ClusterImage clusterImage
     ) {
         // Copy and update the current subscriptions.
         Map<String, Integer> subscribedTopicNames = new HashMap<>(this.subscribedTopicNames);
         maybeUpdateSubscribedTopicNames(subscribedTopicNames, oldMember, newMember);
 
         // Create the topic metadata for each subscribed topic.
         Map<String, TopicMetadata> newSubscriptionMetadata = new HashMap<>(subscribedTopicNames.size());
+
         subscribedTopicNames.forEach((topicName, count) -> {
             TopicImage topicImage = topicsImage.getTopic(topicName);
             if (topicImage != null) {
+                Map<Integer, Set<String>> partitionRackInfo = new HashMap<>();
+
+                topicImage.partitions().forEach((partition, partitionRegistration) -> {
+                    Set<String> racks = new HashSet<>();
+                    for (int replica : Objects.requireNonNull(partitionRegistration.replicas)) {
+                        Optional<String> rackOptional = clusterImage.broker(replica).rack();
+                        // Only add rack if it is available for the broker/replica.
+                        rackOptional.ifPresent(racks::add);
+                    }
+                    partitionRackInfo.put(partition, racks);
```

Comment:
I am thinking about a possible optimization. We could actually keep `numPartitions` and put partition's metadata only if racks is non-empty. That would save memory in clusters without rack awareness. What do you think?

Reply:
yeah I agree